dandi/upload.py

```diff
@@ -573,3 +587,272 @@ def upload_agg(*ignored):
+        )
+    if not re.match(dandiset_identifier_regex, ds_identifier):
+        raise ValueError(
+            f"Dandiset identifier {ds_identifier} does not follow expected "
+            f"convention {dandiset_identifier_regex!r}.  Use "
+            f"'dandi register' to get a legit identifier"
+        )
+    # this is a path not a girder id
+
+    from .pynwb_utils import ignore_benign_pynwb_warnings, get_object_id
+    from .metadata import get_metadata
+    from .validate import validate_file
+    from .utils import find_dandi_files, find_files, path_is_subpath
+    from .support.pyout import naturalsize
+
+    ignore_benign_pynwb_warnings()  # so validate doesn't whine
+
+    #
+    # Treat paths
+    #
+    if not paths:
+        paths = [dandiset.path]
+
+    # Expand and validate all paths -- they should reside within dandiset
+    paths = find_files(".*", paths) if allow_any_path else find_dandi_files(paths)
+    paths = list(map(Path, paths))
+    npaths = len(paths)
+    lgr.info(f"Found {npaths} files to consider")
+    for path in paths:
+        if not (
+            allow_any_path
+            or path.name == dandiset_metadata_file
+            or path.name.endswith(".nwb")
+        ):
+            raise NotImplementedError(
+                f"ATM only .nwb and dandiset.yaml should be in the paths to upload. Got {path}"
+            )
+        if not path_is_subpath(str(path.absolute()), dandiset.path):
+            raise ValueError(f"{path} is not under {dandiset.path}")
+
+    # We will keep a shared set of "being processed" paths so
+    # we could limit the number of them until
+    #   https://github.com/pyout/pyout/issues/87
+    # properly addressed
+    process_paths = set()
+    from collections import defaultdict
+
+    uploaded_paths = defaultdict(lambda: {"size": 0, "errors": []})
+
+    def skip_file(msg):
+        return {"status": "skipped", "message": str(msg)}
+
+    # TODO: we might want to always yield a full record so no field is not
+    # provided to pyout to cause it to halt
+    def process_path(path, relpath):
+        """
+
+        Parameters
+        ----------
+        path: Path
+          Non Pure (OS specific) Path
+        relpath:
+          For location on server.  Will be cast to PurePosixPath
+
+        Yields
+        ------
+        dict
+          Records for pyout
+        """
+        # Ensure consistent types
+        path = Path(path)
+        relpath = PurePosixPath(relpath)
+        try:
+            try:
+                path_stat = path.stat()
+                yield {"size": path_stat.st_size}
+            except FileNotFoundError:
+                yield skip_file("ERROR: File not found")
+                return
+            except Exception as exc:
+                # without limiting [:50] it might cause some pyout indigestion
+                yield skip_file("ERROR: %s" % str(exc)[:50])
+                return
+
```

Comment:
here we are to use `dandisets_versions_assets_paths` end point to mimic prior logic to decide on when to re-upload the file/asset if it already exists (sorry if I misguided in a quick reply before when you asked about girder Item's - those indeed we should not care about creating/delete).  We should not reupload if file/asset is already there and identical to local one. Now we will just make decision solely based on digest (sha256) on either file needs reupload. 

Later we might want to add a mode to only update metadata (thus minting a new asset) for an existing file.